PresentData/PresentMonTraceConsumer.cpp

```diff
@@ -471,6 +496,10 @@ void PMTraceConsumer::HandleDxgkSubmitPresentHistoryEventArgs(
     eventIter->second->FinalState = PresentResult::Unknown;
     eventIter->second->TokenPtr = token;
 
+    // TODO: What happens if there already is an item in mDxgKrnlPresentHistoryTokens?

```

Comment:
This token is a kernel memory address, and so should be unique unless it was deallocated/reallocated with the same pointer, while we had a frame stuck in this map. That should never happen, so an assert seems like a good idea here.

Reply:
Heads up that this assert has triggered in selfhost:

https://www.dropbox.com/s/pc2b5rdifcbsrg9/TokenPtrDump.zip?dl=0

```
Prexisting: [Type: std::shared_ptr<PresentEvent>]
    [<Raw View>]     [Type: std::shared_ptr<PresentEvent>]
    [ptr]            : 0x1bb7e14d4b0 [Type: PresentEvent *]
    [control block]  [Type: std::_Ref_count_obj2<PresentEvent> (derived from std::_Ref_count_base)]
    [+0x000] QpcTime          : 0x99d3f4ecb0 [Type: unsigned __int64]
    [+0x008] ProcessId        : 0x76a4 [Type: unsigned int]
    [+0x00c] ThreadId         : 0x7a18 [Type: unsigned int]
    [+0x010] TimeTaken        : 0x25367 [Type: unsigned __int64]
    [+0x018] ReadyTime        : 0x0 [Type: unsigned __int64]
    [+0x020] ScreenTime       : 0x0 [Type: unsigned __int64]
    [+0x028] SwapChainAddress : 0x89b3140 [Type: unsigned __int64]
    [+0x030] SyncInterval     : -1 [Type: int]
    [+0x034] PresentFlags     : 0x0 [Type: unsigned int]
    [+0x038] Hwnd             : 0x381342 [Type: unsigned __int64]
    [+0x040] TokenPtr         : 0xffffe482a95d68c0 [Type: unsigned __int64]
    [+0x048] QueueSubmitSequence : 0x15ca4b [Type: unsigned int]
    [+0x04c] Runtime          : D3D9 (1) [Type: Runtime]
    [+0x050] PresentMode      : Composed_Copy_GPU_GDI (5) [Type: PresentMode]
    [+0x054] FinalState       : Unknown (0) [Type: PresentResult]
    [+0x058] DestWidth        : 0x0 [Type: unsigned int]
    [+0x05c] DestHeight       : 0x0 [Type: unsigned int]
    [+0x060] CompositionSurfaceLuid : 0x0 [Type: unsigned __int64]
    [+0x068] DriverBatchThreadId : 0x53d0 [Type: unsigned int]
    [+0x06c] SupportsTearing  : false [Type: bool]
    [+0x06d] MMIO             : false [Type: bool]
    [+0x06e] SeenDxgkPresent  : true [Type: bool]
    [+0x06f] SeenWin32KEvents : false [Type: bool]
    [+0x070] DwmNotified      : false [Type: bool]
    [+0x071] Completed        : false [Type: bool]
    [+0x072] IsLost           : false [Type: bool]
    [+0x074] mAllPresentsTrackingIndex : 0x1939 [Type: unsigned int]
    [+0x078] DxgKrnlHContext  : 0x0 [Type: unsigned __int64]
    [+0x080] Win32KPresentCount : 0x0 [Type: unsigned __int64]
    [+0x088] Win32KBindId     : 0x0 [Type: unsigned __int64]
    [+0x090] LegacyBlitTokenData : 0x0 [Type: unsigned __int64]
    [+0x098] DependentPresents : { size=0x0 } [Type: std::deque<std::shared_ptr<PresentEvent>,std::allocator<std::shared_ptr<PresentEvent> > >]
    [+0x0c0] PresentInDwmWaitingStruct : false [Type: bool]
```

New present: [Type: std::shared_ptr<PresentEvent>]
```
    [<Raw View>]     [Type: std::shared_ptr<PresentEvent>]
    [ptr]            : 0x1bb7e14fe50 [Type: PresentEvent *]
    [control block]  [Type: std::_Ref_count_obj2<PresentEvent> (derived from std::_Ref_count_base)]
    [+0x000] QpcTime          : 0x99d3f6ffca [Type: unsigned __int64]
    [+0x008] ProcessId        : 0x77ec [Type: unsigned int]
    [+0x00c] ThreadId         : 0xecc [Type: unsigned int]
    [+0x010] TimeTaken        : 0x10103db [Type: unsigned __int64]
    [+0x018] ReadyTime        : 0x0 [Type: unsigned __int64]
    [+0x020] ScreenTime       : 0x0 [Type: unsigned __int64]
    [+0x028] SwapChainAddress : 0x2021393d1f0 [Type: unsigned __int64]
    [+0x030] SyncInterval     : 1 [Type: int]
    [+0x034] PresentFlags     : 0x0 [Type: unsigned int]
    [+0x038] Hwnd             : 0x0 [Type: unsigned __int64]
    [+0x040] TokenPtr         : 0xffffe482a95d68c0 [Type: unsigned __int64]
    [+0x048] QueueSubmitSequence : 0x0 [Type: unsigned int]
    [+0x04c] Runtime          : DXGI (0) [Type: Runtime]
    [+0x050] PresentMode      : Composed_Flip (4) [Type: PresentMode]
    [+0x054] FinalState       : Unknown (0) [Type: PresentResult]
    [+0x058] DestWidth        : 0x384 [Type: unsigned int]
    [+0x05c] DestHeight       : 0x4ae [Type: unsigned int]
    [+0x060] CompositionSurfaceLuid : 0x9fceef2 [Type: unsigned __int64]
    [+0x068] DriverBatchThreadId : 0x0 [Type: unsigned int]
    [+0x06c] SupportsTearing  : false [Type: bool]
    [+0x06d] MMIO             : false [Type: bool]
    [+0x06e] SeenDxgkPresent  : false [Type: bool]
    [+0x06f] SeenWin32KEvents : true [Type: bool]
    [+0x070] DwmNotified      : false [Type: bool]
    [+0x071] Completed        : false [Type: bool]
    [+0x072] IsLost           : false [Type: bool]
    [+0x074] mAllPresentsTrackingIndex : 0x193a [Type: unsigned int]
    [+0x078] DxgKrnlHContext  : 0x0 [Type: unsigned __int64]
    [+0x080] Win32KPresentCount : 0x81e27 [Type: unsigned __int64]
    [+0x088] Win32KBindId     : 0x3 [Type: unsigned __int64]
    [+0x090] LegacyBlitTokenData : 0x0 [Type: unsigned __int64]
    [+0x098] DependentPresents : { size=0x0 } [Type: std::deque<std::shared_ptr<PresentEvent>,std::allocator<std::shared_ptr<PresentEvent> > >]
    [+0x0c0] PresentInDwmWaitingStruct : false [Type: bool]
```
PreExisting:
mAllPresentsTrackingIndex : 0x1939
30372 = Baba is You.exe

New:
mAllPresentsTrackingIndex : 0x193a
30700 = Solitaire.exe

Current mAllPresentsNextIndex: 0x193c

These are windowed blit and windowed flip apps on the same monitor in a multi-monitor scenario.

Neither has been marked as lost. Perhaps the old present got deallocated, stuck, and allocated again? If this is expected, then the assert might not be very useful if selfhost can hit it frequently.

Reply:
Screenshot:
![image](https://user-images.githubusercontent.com/1754267/98891630-76443880-2453-11eb-8d10-b6110eec7532.png)

Reply:
Also @jdm3 unfortunately I don't have a good way of being notified if you replied to a thread but didn't unresolve the conversation :( let me know if there are comments I accidentally ignored because they are in a resolved conversation.

Reply:
That's... odd. I have no idea how that would happen.

Also, there's no reason to put Win8+ flip model presents in that map, we only need Win7 flip, blt, and DComp (composition atlas) presents there. It probably doesn't hurt anything to put them there, and it might be extra complexity to avoid putting them in there, just wanted to call it out.